drivers/counter/counter_mcux_lpc_rtc.c

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright (c) 2021, NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nxp_lpc_rtc
+
+#include <drivers/counter.h>
+#include <fsl_rtc.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(mcux_rtc, CONFIG_COUNTER_LOG_LEVEL);
+
+struct mcux_lpc_rtc_data {
+	counter_alarm_callback_t alarm_callback;
+	counter_top_callback_t top_callback;
+	void *alarm_user_data;
+	void *top_user_data;
+};
+
+struct mcux_lpc_rtc_config {
+	struct counter_config_info info;
+	RTC_Type *base;
+	void (*irq_config_func)(const struct device *dev);
+};
+
+static int mcux_lpc_rtc_start(const struct device *dev)
+{
+	const struct counter_config_info *info = dev->config;
+	const struct mcux_lpc_rtc_config *config =
+		CONTAINER_OF(info, struct mcux_lpc_rtc_config, info);
+
+	RTC_StartTimer(config->base);
+
+	return 0;
+}
+
+static int mcux_lpc_rtc_stop(const struct device *dev)
+{
+	const struct counter_config_info *info = dev->config;
+	const struct mcux_lpc_rtc_config *config =
+		CONTAINER_OF(info, struct mcux_lpc_rtc_config, info);
+
+	RTC_StopTimer(config->base);
+
+	/* clear out any set alarms */
+	RTC_SetSecondsTimerMatch(config->base, 0);
+
+	return 0;
+}
+
+static uint32_t mcux_lpc_rtc_read(const struct device *dev)
+{
+	const struct counter_config_info *info = dev->config;
+	const struct mcux_lpc_rtc_config *config =
+		CONTAINER_OF(info, struct mcux_lpc_rtc_config, info);
+
+	uint32_t ticks = RTC_GetSecondsTimerCount(config->base);
+
+	return ticks;
+}
+
+static int mcux_lpc_rtc_get_value(const struct device *dev, uint32_t *ticks)
+{
+	*ticks = mcux_lpc_rtc_read(dev);
+	return 0;
+}
+
+static int mcux_lpc_rtc_set_alarm(const struct device *dev, uint8_t chan_id,
+			      const struct counter_alarm_cfg *alarm_cfg)
+{
+	const struct counter_config_info *info = dev->config;
+	const struct mcux_lpc_rtc_config *config =
+		CONTAINER_OF(info, struct mcux_lpc_rtc_config, info);
+	struct mcux_lpc_rtc_data *data = dev->data;
+
+	uint32_t ticks = alarm_cfg->ticks;
+	uint32_t current = mcux_lpc_rtc_read(dev);
+
+	LOG_DBG("Current time is %d ticks", current);
+
+	if (chan_id != 0U) {
+		LOG_ERR("Invalid channel id");
+		return -EINVAL;
+	}
+
+	if (data->alarm_callback != NULL) {
+		return -EBUSY;
+	}
+
+	if ((alarm_cfg->flags & COUNTER_ALARM_CFG_ABSOLUTE) == 0) {
+		ticks += current;
+	}
+
+	if (ticks < current) {
+		LOG_ERR("Alarm cannot be earlier than current time");
+		return -EINVAL;
+	}
+
+	data->alarm_callback = alarm_cfg->callback;
+	data->alarm_user_data = alarm_cfg->user_data;
+
+	RTC_SetSecondsTimerMatch(config->base, ticks);
```

Comment:
there is a potential race condition here, if alarm got cancelled then only callback is set to null, if it expires after setting new alarm_callback but before setting new alarm value then it will result in premature expiration of a new alarm.